### PR TITLE
Cognito domain creation and expose OAuth to frontend config

### PR DESCRIPTION
**Pull request by awsluja**
*Issue #, if available:*

*Description of changes:*
This change enables the creation of a Cognito domain given the specified 'domainPrefix'.
It allows users to create, update, or delete the domain after initial deployments (with a slight workaround).

Also, exposes the OAuth settings to the frontend config.
- social providers 
- oauth (domain, scope, signin/out urls, response type)

UX:
image

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: packages/auth-construct/src/construct.ts [hunk 1]
```diff
@@ -712,18 +756,41 @@ export class AmplifyAuth
...
+          this.oAuthSettings.scopes?.map((s) => s.scopeName) ?? []
+        );
+        output.oauthRedirectSignIn = this.oAuthSettings.callbackUrls
+          ? this.oAuthSettings.callbackUrls.join(',')
+          : '';
+        output.oauthRedirectSignOut = this.oAuthSettings.logoutUrls
+          ? this.oAuthSettings.logoutUrls.join(',')
+          : '';
+        output.webClientId = this.resources.userPoolClient.userPoolClientId;
+        output.oauthResponseType = 'code';
```
Comment (awsluja) [inline packages/auth-construct/src/construct.ts:792]: note: 'code' seems to be the default for hosted ui (old cli), when would this not be the case?

File: packages/auth-construct/src/types.ts [hunk 2]
```diff
@@ -199,11 +199,19 @@ export type ExternalProviderOptions = {
   /**
    * List of allowed redirect URLs for the identity providers.
    */
-  callbackUrls?: string[];
+  callbackUrls: string[];
```
Comment (edwardfoyle) [inline packages/auth-construct/src/types.ts:202]: not strictly part of this PR, but this should either be `redirectUrls` to match the comment, or change the comment to say "callback URLs"

File: packages/auth-construct/src/construct.ts (hunk 1 above)
Comment (edwardfoyle) [reply packages/auth-construct/src/construct.ts:792]: is this still an open question?

**Review by edwardfoyle — DISMISSED** [review]

File: packages/auth-construct/src/construct.ts (hunk 1 above)
Comment (awsluja) [reply packages/auth-construct/src/construct.ts:792]: not anymore; 'token' is considered legacy and less secure than 'code' so we don't have a need for 'token' at the moment

File: packages/auth-construct/src/types.ts (hunk 2 above)
Comment (awsluja) [reply packages/auth-construct/src/types.ts:202]: image
I'm not sure why but CDK uses the names interchangeably here; I'm not sure if this is intentional

File: packages/auth-construct/src/construct.ts [hunk 3]
```diff
@@ -108,32 +121,56 @@ export class AmplifyAuth
...
     // UserPool - Identity Providers
-    const providerSetupResult = this.setupIdentityProviders(
+    this.providerSetupResult = this.setupIdentityProviders(
       this.userPool,
       props.loginWith
     );
-    this.oauthMappings = providerSetupResult.oauthMappings;
 
-    // UserPool Client
+    this.domainPrefix = props.loginWith.externalProviders?.domainPrefix ?? '';
```
Comment (sobolk) [inline packages/auth-construct/src/construct.ts:129]: why do we need `?? ''` if we check it in `if` below ?
is `this.domainPrefix` type correct?

File: packages/auth-construct/src/construct.ts [hunk 4]
```diff
@@ -108,32 +121,56 @@ export class AmplifyAuth
...
+    if (
+      this.providerSetupResult.providersList.length > 0 &&
+      this.domainPrefix
```
Comment (sobolk) [inline packages/auth-construct/src/construct.ts:132]: should we check providers list here?
if I  explicitly provide only `domainPrefix` in inputs I'd be surprised to see it not created.
alternatively we should validate and throw if providing domain prefix doesn't make sense if providers list is empty.

Comment (awsluja) [reply packages/auth-construct/src/construct.ts]: I went with the latter, throwing an error if domainPrefix is specified with an empty providers list

File: packages/auth-construct/src/construct.ts (hunk 3 above)
Comment (awsluja) [reply packages/auth-construct/src/construct.ts]: good catch, i updated the type

**Review by edwardfoyle — APPROVED** [review]

**Review by 0618 — APPROVED** [review]

**Pull request merged**